test/limitOrderBook/limitOrderBook.fillOrder.test.ts

```diff
@@ -0,0 +1,575 @@
+            // long 0.1 ETH at $3000 with $300
+            const limitOrder = {
+                orderType: fixture.orderTypeLimitOrder,
+                salt: 1,
+                trader: trader.address,
+                baseToken: baseToken.address,
+                isBaseToQuote: false,
+                isExactInput: true,
+                amount: parseEther("300"),
+                oppositeAmountBound: parseEther("0.1"),
+                deadline: ethers.constants.MaxUint256,
+                reduceOnly: true,
+                roundIdWhenCreated: parseEther("0").toString(),
+                triggerPrice: parseEther("0").toString(),
+            }
+
+            // sign limit order
+            const signature = await getSignature(fixture, limitOrder, trader)
+
+            await expect(
+                limitOrderBook.connect(keeper).fillLimitOrder(limitOrder, signature, parseEther("0")),
+            ).to.be.revertedWith("LOB_NRO")
+        })
+
+        it("force error, when order does not satisfy reduceOnly, create a reverse position", async () => {
+            // short 0.2 ETH at $2800 with $560
+            const limitOrder = {
+                orderType: fixture.orderTypeLimitOrder,
+                salt: 1,
+                trader: trader.address,
+                baseToken: baseToken.address,
+                isBaseToQuote: true,
+                isExactInput: true,
+                amount: parseEther("0.2").toString(),
+                oppositeAmountBound: parseEther("560").toString(),
+                deadline: ethers.constants.MaxUint256.toString(),
+                reduceOnly: true,
+                roundIdWhenCreated: parseEther("0").toString(),
+                triggerPrice: parseEther("0").toString(),
+            }
+
+            // sign limit order
+            const signature = await getSignature(fixture, limitOrder, trader)
+
+            await expect(
+                limitOrderBook.connect(keeper).fillLimitOrder(limitOrder, signature, parseEther("0")),
+            ).to.be.revertedWith("LOB_NRO")
+        })
+    })
+
+    it("cancel order successfully", async () => {
+        // long 0.1 ETH at $3000 with $300
+        const limitOrder = {
+            orderType: fixture.orderTypeLimitOrder,
+            salt: 1,
+            trader: trader.address,
+            baseToken: baseToken.address,
+            isBaseToQuote: false,
+            isExactInput: true,
+            amount: parseEther("300"),
+            oppositeAmountBound: parseEther("0.1"),
+            deadline: ethers.constants.MaxUint256,
+            reduceOnly: false,
+            roundIdWhenCreated: parseEther("0").toString(),
+            triggerPrice: parseEther("0").toString(),
+        }
+
+        await expect(await limitOrderBook.connect(trader).cancelLimitOrder(limitOrder)).to.emit(
+            limitOrderBook,
+            "LimitOrderCancelled",
+        )
+    })
+
```

Comment:
what would happen if we cancel an order twice? We might need a test case to clarify the spec.